### PR TITLE
chore: Add redux to I3S Explorer app

### PR DESCRIPTION
**Pull request by MaximKYZ**
Redux is added into the App
Created redux store and slices for selecting and dispatching colorsByAttributes and async fetching flattened layers and sublayers
A number of tests have been reworked to support redux store

### CODE REVIEW

File: src/redux/colors-by-attribute-slice.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { createSlice } from "@reduxjs/toolkit";
+import { ColorsByAttribute } from "../types";
+import { RootState } from "./store";
+
+// Define a type for the slice state
+interface ColorsByAttributeState {
+    value: ColorsByAttribute | null;
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/colors-by-attribute-slice.tsx:7]: Done

File: src/redux/colors-by-attribute-slice.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+            state.value = action.payload;
+        },
+    },
+});
+
+export const selectColorsByAttribute = (state: RootState) => state.colorsByAttribute.value;
+
+export const { setColorsByAttrubute } = colorsByAttributeSlice.actions;
+
+export default colorsByAttributeSlice.reducer;
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/colors-by-attribute-slice.tsx:29]: Done

File: src/redux/flattened-sublayers-slice.tsx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        builder
+          .addCase(getFlattenedSublayers.fulfilled, (state, action) => {
+            if (action.payload.layerCounter > state.layerCounter) {
+              return {layers: action.payload.layers, layerCounter: action.payload.layerCounter, sublayers: action.payload.sublayers};
+            }
+          })
+      }
+});
+
+export const getFlattenedSublayers = createAsyncThunk('getFlattenedSublayers', async ({ tilesetsData, currentLayer, buildingExplorerOpened }: { tilesetsData: tilesetsDataType[], currentLayer: number, buildingExplorerOpened: boolean }) => {
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/flattened-sublayers-slice.tsx:48]: Done

File: src/redux/flattened-sublayers-slice.tsx [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    for (const data of tilesetsData) {
+        if (!data.hasChildren) {
+            promises.push(getLayersAndSublayers(data, buildingExplorerOpened));
+        }
+    }
+
+    const layers = await Promise.all(promises).then((results) => {
+      return results;
+    })
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/flattened-sublayers-slice.tsx:59]: Done

File: src/redux/flattened-sublayers-slice.tsx [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+      },
+      setBuildingSubLayers: (state, action) => {
+        state.sublayers = action.payload;
+      },
+    },
+    extraReducers(builder) {
+        builder
+          .addCase(getFlattenedSublayers.fulfilled, (state, action) => {
+            if (action.payload.layerCounter > state.layerCounter) {
+              return {layers: action.payload.layers, layerCounter: action.payload.layerCounter, sublayers: action.payload.sublayers};
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/flattened-sublayers-slice.tsx:42]: Done

File: src/redux/hooks.tsx [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";
+import type { RootState, AppDispatch } from "./store";
+
+// Use throughout your app instead of plain `useDispatch` and `useSelector`
+export const useAppDispatch: () => AppDispatch = useDispatch;
+export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/hooks.tsx:6]: Done

File: src/redux/store.tsx [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
+  return configureStore({
+    reducer: rootReducer,
+    preloadedState
+  });
+}
+
+export type RootState = ReturnType<typeof rootReducer>;
+export type AppStore = ReturnType<typeof setupStore>;
+export type AppDispatch = AppStore['dispatch'];
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/store.tsx:20]: Done

File: src/utils/testing-utils/render-with-provider.tsx [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+export function renderWithProvider(
+  ui: React.ReactElement,
+  store: AppStore
+) {
+// eslint-disable-next-line @typescript-eslint/ban-types
+function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
+  return <Provider store={store}>{children}</Provider>;
+}
+  return { store, ...render(ui, { wrapper: Wrapper }) };
+}
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/utils/testing-utils/render-with-provider.tsx:15]: Done

File: src/types.ts [hunk 9]
```diff
@@ -354,3 +354,11 @@ export type TileSelectedColor = {
   g: number;
   b: number;
 };
+
+export type tilesetsDataType = {
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/types.ts:358]: Done

File: src/pages/debug-app/debug-app.tsx [hunk 10]
```diff
@@ -267,8 +236,7 @@ export const DebugApp = () => {
       });
     }
 
-    fetchFlattenedSublayers(tilesetsData, fetchSublayersCounter.current);
-    setSublayers([]);
+    dispatch(getFlattenedSublayers({tilesetsData: tilesetsData, currentLayer: fetchSublayersCounter.current, buildingExplorerOpened: buildingExplorerOpened}));
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/pages/debug-app/debug-app.tsx:239]: Done

File: src/redux/flattened-sublayers-slice.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/flattened-sublayers-slice.tsx]: Done

File: src/redux/flattened-sublayers-slice.tsx [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+  layers: [],
+  layerCounter: 0,
+  sublayers: []
+};
+
+const flattenedSublayersSlice = createSlice({
+    name: 'flattenedSublayers',
+    initialState,
+    reducers: {
+      setFlattenedSublayers: (state, action) => {
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/flattened-sublayers-slice.tsx]: Done

File: src/redux/flattened-sublayers-slice.tsx [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+          state.layers[action.payload.index].visibility = action.payload.visibility;
+        }
+      },
+      setBuildingSubLayers: (state, action) => {
+        state.sublayers = action.payload;
+      },
+    },
+    extraReducers(builder) {
+        builder
+          .addCase(getFlattenedSublayers.fulfilled, (state, action) => {
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/flattened-sublayers-slice.tsx]: Done

File: src/redux/flattened-sublayers-slice.tsx [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    const layers = await Promise.all(promises).then((results) => {
+      return results;
+    })
+
+    return {layers: layers[0].layers.flat(), layerCounter: currentLayer, sublayers: layers[0].sublayers};
+});
+
+const getLayersAndSublayers = async (tilesetData: tilesetsDataType, buildingExplorerOpened: boolean) => {
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/flattened-sublayers-slice.tsx]: Done

File: src/redux/flattened-sublayers-slice.tsx [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+  }
+};
+
+export const selectLayers = (state: RootState) => state.flattenedSublayers.layers;
+export const selectSublayers = (state: RootState) => state.flattenedSublayers.sublayers;
+export const selectLayerCounter = (state: RootState) => state.flattenedSublayers.layerCounter;
+
+export const { setFlattenedSublayers, updateLayerVisibility, setBuildingSubLayers } = flattenedSublayersSlice.actions;
+
+export default flattenedSublayersSlice.reducer;
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/flattened-sublayers-slice.tsx]: Done

File: src/redux/flattened-sublayers-slice.tsx [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        type: tilesetData.type,
+      },
+    ], sublayers: []};
+    return result;
+  }
+};
+
+export const selectLayers = (state: RootState) => state.flattenedSublayers.layers;
+export const selectSublayers = (state: RootState) => state.flattenedSublayers.sublayers;
+export const selectLayerCounter = (state: RootState) => state.flattenedSublayers.layerCounter;
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/flattened-sublayers-slice.tsx]: Done

File: src/pages/debug-app/debug-app.tsx [hunk 16]
```diff
@@ -163,15 +169,17 @@ export const DebugApp = () => {
   const [examples, setExamples] = useState<LayerExample[]>(EXAMPLES);
   const [activeLayers, setActiveLayers] = useState<LayerExample[]>([]);
   const [viewState, setViewState] = useState<ViewStateSet>(INITIAL_VIEW_STATE);
-  const fetchSublayersCounter = useRef<number>(0);
+  const fetchSublayersCounter = useRef<number>(
```
Comment (belom88) [inline src/pages/debug-app/debug-app.tsx:172]: I believe we can serve this logic with "counter" inside the redux

File: src/types.ts [hunk 17]
```diff
@@ -347,10 +347,18 @@ export type TileInfo = {
...
-}
+};
 
 export type TileSelectedColor = {
   r: number;
   g: number;
   b: number;
 };
+
+export type tilesetsData = {
```
Comment (belom88) [inline src/types.ts:358]:
```suggestion
export type TilesetMetadata = {
```

File: src/utils/testing-utils/render-with-theme.tsx [hunk 18]
```diff
@@ -32,3 +32,16 @@ const theme = {
...
+  store: AppStore
+) {
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
+    return <Provider store={store}>
+      <ThemeProvider theme={theme}>{children}</ThemeProvider>
+    </Provider>;
+  }
+  return { store, ...render(ui, { wrapper: Wrapper }) };
+}
```
Comment (belom88) [inline src/utils/testing-utils/render-with-theme.tsx:47]: new line in the end of the file

File: src/redux/colors-by-attribute-slice.tsx [hunk 19]
```diff
@@ -0,0 +1,34 @@
+import { PayloadAction, createSlice } from "@reduxjs/toolkit";
+import { ColorsByAttribute } from "../types";
+import { RootState } from "./store";
+
+// Define a type for the slice state
+interface ColorsByAttributeState {
+  value: ColorsByAttribute | null;
```
Comment (belom88) [inline src/redux/colors-by-attribute-slice.tsx:7]: Could you add TSDoc comments for state items?

File: src/redux/flattened-sublayers-slice.tsx [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+  tilesetsData,
+} from "../types";
+import { I3SBuildingSceneLayerLoader } from "@loaders.gl/i3s";
+import { buildSublayersTree } from "../utils/sublayers";
+import { load } from "@loaders.gl/core";
+import { RootState } from "./store";
+
+// Define a type for the slice state
+interface flattenedSublayersState {
+  layers: BuildingSceneSublayerExtended[];
```
Comment (belom88) [inline src/redux/flattened-sublayers-slice.tsx:15]: Ditto TSDoc comments

**Review by belom88 — CHANGES_REQUESTED** [review]

File: src/redux/drag-mode-slice.tsx [hunk 21]
```diff
@@ -0,0 +1,26 @@
+import { PayloadAction, createSlice } from "@reduxjs/toolkit";
```
[the remark this answers is not in the document]
Comment (MaximKYZ) [reply src/redux/drag-mode-slice.tsx:1]: Done

File: src/types.ts (hunk 17 above)
Comment (MaximKYZ) [reply src/types.ts:358]: Done

File: src/utils/testing-utils/render-with-theme.tsx (hunk 18 above)
Comment (MaximKYZ) [reply src/utils/testing-utils/render-with-theme.tsx:47]: Done

File: src/pages/debug-app/debug-app.tsx (hunk 16 above)
Comment (MaximKYZ) [reply src/pages/debug-app/debug-app.tsx:172]: Done

File: src/redux/flattened-sublayers-slice.tsx (hunk 20 above)
Comment (MaximKYZ) [reply src/redux/flattened-sublayers-slice.tsx:15]: Done

File: src/redux/colors-by-attribute-slice.tsx (hunk 19 above)
Comment (MaximKYZ) [reply src/redux/colors-by-attribute-slice.tsx:7]: Done

**Review by belom88 — APPROVED** [review]

**Pull request merged**